src/Symfony/Component/Runtime/SymfonyRuntime.php

```diff
@@ -0,0 +1,198 @@
+            }
+        } catch (\InvalidArgumentException $e) {
+            if (!$parameter->isOptional()) {
+                throw $e;
+            }
+        }
+
+        return [$main, $arguments];
+    }
+
+    public function getHandler(object $result): \Closure
+    {
+        if ($result instanceof HttpKernelInterface) {
+            $request = $this->request ?? $this->request = Request::createFromGlobals();
+
+            return static function (HttpKernelInterface $kernel) use ($request): int {
+                $response = $kernel->handle($request);
+                $response->send();
+
+                if ($kernel instanceof TerminableInterface) {
+                    $kernel->terminate($request, $response);
+                }
+
+                return 0;
+            };
+        }
+
+        if ($result instanceof Response) {
+            return static function (Response $response): int {
+                $response->send();
+
+                return 0;
+            };
+        }
+
+        if ($result instanceof Command) {
+            return function (Command $command): int {
+                $application = $this->application ?? $this->application = new Application();
+                $application->setName($command->getName() ?: $application->getName());
+
+                if (!$command->getName() || !$application->has($command->getName())) {
+                    $command->setName($_SERVER['argv'][0]);
+                    $application->add($command);
+                }
+
+                $application->setDefaultCommand($command->getName(), true);
+
+                return $this->getHandler($application)($application);
+            };
+        }
+
+        if ($result instanceof Application) {
+            $input = $this->getInput();
+
+            return static function (Application $application) use ($input): int {
+                if (!\in_array(\PHP_SAPI, ['cli', 'phpdbg', 'embed'], true)) {
+                    echo 'Warning: The console should be invoked via the CLI version of PHP, not the '.\PHP_SAPI.' SAPI'.\PHP_EOL;
+                }
+
+                set_time_limit(0);
+
+                $definition = $application->getDefinition();
+
+                if (!$definition->hasOption('env')) {
+                    $definition->addOption(new InputOption('--env', '-e', InputOption::VALUE_REQUIRED, 'The Environment name.', $_SERVER['APP_ENV']));
+                }
+
+                if (!$definition->hasOption('no-debug')) {
+                    $definition->addOption(new InputOption('--no-debug', null, InputOption::VALUE_NONE, 'Switches off debug mode.'));
+                }
+
+                return $application->run($input);
+            };
+        }
+
+        throw new \LogicException(sprintf('Cannot handle return value of type "%s".', \get_class($result)));
+    }
+
+    protected function getArgument(\ReflectionParameter $parameter)
+    {
+        switch ($parameter->getType()->getName()) {
+            case 'array':
```

Comment:
I'd better suggest to create a separate class to `autowire` such things. Mapping any `context` array to $_SERVER will later make it harder to implement new Runtimes keeping witout this behaviour or rewriting mains. I suggest keep any primitive types to be only in userland implementation to avoid "vendor-locks". Binding symfony classes in symfony runtimes looks fully valid, but seizing primitives would cause problems in the future IMO

Something like this would make things a bit clear, I think
```php
case ServerContext::class // maybe some kind of ParameterBag successor?
  return $this->context ?? $this->context = ServerContext::createFromGlobals();
```

Reply:
> create a separate class to autowire such things.

I thought about this and decided to NOT go this way: a dedicated class would mean coupling to the package that declares it. THAT would be "vendor-lockin". That's why I didn't do it.

I think this `$context` array is the most interoperable thing that exists in the PHP engine. Incompatible runtimes would throw a proper exception to make the situation actionable I suppose.

Another reason why I decided this way is because this can be reconsidered later on, when we have more experience with the system.

Reply:
> THAT would be "vendor-lockin"

> Binding symfony classes in symfony runtimes looks fully valid

This couldn't be lockin since anyone who don't want to use the symfony classes just doesn't declare it in his Main signature. Same stands for symfony Request, Response and other usefull classes. Currently It's possible to write a complex runtime which covers a lot of cases even my old ZF2 applications, and there will be no lockin if SymfonyRuntime declare it's own `$_SERVER` wrapper in the same way it did with Requets class for $_GET, $_POST, etc.

The difference between `array $context` and `ServerContext $context` is that first one works with some kind of magic declared in symfony runtime (and different runtime can fill it differently causing bugs) and second one is the explicit wish of the `Main`'s writer to obtain ServerContext instance as the argument

Reply:
OK, I get the point. I just don't know in which package `ServerContext` would fit. Maybe in a future (and hypothetic) `symfony/runtime-contracts`. Let's resolve this later on if you don't mind.